File: dnn-providers/hip-kernel-provider/rocke/platform/Python/rocke/helpers/scan.py
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT

"""Cooperative block-wide scan + histogram helpers.

CK Tile's MoE-sort and similar bucket-style kernels share three
patterns that aren't covered by the existing
:func:`rocke.helpers.block_lds_reduce`:

* **Block histogram**: every lane contributes one or more ``(key, +1)``
  pairs; the kernel materialises an ``(N,)`` count array in LDS.
  Implemented as a zero-LDS prologue + cooperative
  :meth:`IRBuilder.lds_atomic_add` chain + sync.

* **Block exclusive scan**: given an ``(N,)`` LDS array, compute the
  exclusive prefix sum in place. Used to convert per-bucket counts
  into per-bucket *offsets*. Uses the classic Hillis-Steele tree
  with ``log2(N)`` LDS round-trips.

* **Atomic bucket counter**: per-bucket "next free slot" counters
  sitting in LDS (or global) that a scatter step ``atomic_add(1)``s
  to claim the next position. The primitive itself is
  :meth:`IRBuilder.lds_atomic_add` / :meth:`IRBuilder.global_atomic_add`;
  this module just packages the standard initialisation pattern
  (zero the counters, sync, then scatter) as a one-liner.

The helpers operate on LDS allocations the caller owns (so the
caller can re-use the buffer across multiple phases when lifetimes
don't overlap), and they assume ``i32`` keys and counts -- the only
case the MoE-sort family needs today. fp32 variants (for soft
histograms) are a straightforward extension if a future kernel asks.
"""

from __future__ import annotations

from typing import Optional, Sequence

from ..core.ir import I32, IRBuilder, Value


__all__ = [
    "lds_zero_i32",
    "block_histogram_i32",
    "block_exclusive_scan_i32",
    "block_two_level_scan_i32",
]


def lds_zero_i32(
    b: IRBuilder,
    lds_buf: Value,
    *,
    tid: Value,
    block_size: int,
    length: int,
) -> None:
    """Cooperatively zero an ``(length,)`` i32 LDS allocation.

    Each lane writes ``ceil(length / block_size)`` slots; the helper
    issues a ``sync`` at the end so subsequent atomic adds against
    the same buffer see the freshly-cleared state.

    For ``length`` that doesn't divide ``block_size`` cleanly, the
    tail lanes are guarded with an in-bounds predicate so out-of-range
    slots are skipped (LDS allocations are exact-sized, so an OOB
    write would corrupt unrelated state).
    """
    if length <= 0:
        raise ValueError(f"length must be > 0 (got {length})")
    chunks = (length + block_size - 1) // block_size
    c_block = b.const_i32(block_size)
    c_length = b.const_i32(length)
    c_zero = b.const_i32(0)
    for c in range(chunks):
        local = b.add(tid, b.mul(b.const_i32(c), c_block))
        in_bounds = b.cmp_lt(local, c_length)
        with b.scf_if(in_bounds):
            b.smem_store_vN(lds_buf, [local], c_zero, 1)
    b.sync()


def block_histogram_i32(
    b: IRBuilder,
    lds_hist: Value,
    keys: Sequence[Value],
    *,
    tid: Value,
    block_size: int,
    num_bins: int,
    valid_mask: Optional[Sequence[Value]] = None,
) -> None:
    """Accumulate per-lane ``keys`` into an LDS histogram of length
    ``num_bins`` (i32).

    Algorithm:

    1. Cooperative zero of ``lds_hist[0..num_bins)`` via
       :func:`lds_zero_i32`.
    2. For each lane-local ``key in keys``: ``atomic_add(lds_hist[key], 1)``.
       The caller chooses how to chunk the keys -- ``len(keys)``
       elements per lane is the typical layout (e.g. ``topk`` for
       MoE-sort, ``vec`` for vectorised histogram passes).
    3. ``sync`` so the histogram is globally visible before any
       follow-up scan.

    ``valid_mask`` is an optional list of i1 predicates (same length
    as ``keys``); when supplied, masked-off entries are skipped. This
    is the canonical "partial last tile" guard.

    ``num_bins`` must equal the LDS allocation's length; this is a
    compile-time fact so no runtime check is emitted.
    """
    if valid_mask is not None and len(valid_mask) != len(keys):
        raise ValueError(
            f"valid_mask length {len(valid_mask)} != keys length {len(keys)}"
        )

    lds_zero_i32(b, lds_hist, tid=tid, block_size=block_size, length=num_bins)

    c_one = b.const_i32(1)
    c_bins = b.const_i32(num_bins)
    for i, key in enumerate(keys):
        if valid_mask is not None:
            in_range = b.land(
                valid_mask[i],
                b.land(b.cmp_ge(key, b.const_i32(0)), b.cmp_lt(key, c_bins)),
            )
        else:
            in_range = b.land(b.cmp_ge(key, b.const_i32(0)), b.cmp_lt(key, c_bins))
        with b.scf_if(in_range):
            b.lds_atomic_add(lds_hist, [key], c_one)
    b.sync()


def block_exclusive_scan_i32(
    b: IRBuilder,
    lds_buf: Value,
    *,
    tid: Value,
    block_size: int,
    length: int,
) -> None:
    """In-place exclusive prefix-sum over ``lds_buf[0..length)`` (i32).

    Hillis-Steele scan with a temporary LDS buffer? No -- we use a
    ping-pong over the *same* buffer with a write-on-other-half
    invariant:

    .. code-block:: text

        for stride in 1, 2, 4, ..., length // 2:
            new_val = (lane >= stride) ? old[lane] + old[lane - stride] : old[lane]
            sync; lds[lane] = new_val; sync

    then shift right by 1 to make the scan exclusive (slot 0 becomes 0,
    slot i becomes sum of [0..i)).

    The kernel must call this with ``length <= block_size`` so every
    lane handles exactly one slot (or none, when ``length < block_size``).
    For larger ``length`` (multi-block scans) the caller should pre-
    reduce per-warp and chain block-level scans -- a future extension.
    """
    if length <= 0:
        raise ValueError(f"length must be > 0 (got {length})")
    if length > block_size:
        raise ValueError(
            f"length {length} > block_size {block_size}; "
            "multi-pass scans not implemented yet"
        )

    c_length = b.const_i32(length)
    in_bounds = b.cmp_lt(tid, c_length)

    # Inclusive Hillis-Steele scan. Both the self-load and the
    # left-neighbour load happen unconditionally on every lane; we
    # clamp the indices to slot 0 for lanes that shouldn't read so
    # the LDS read is always in-bounds. The masked write below keeps
    # those lanes from contributing to the buffer.
    stride = 1
    while stride < length:
        c_stride = b.const_i32(stride)
        do_add = b.land(in_bounds, b.cmp_ge(tid, c_stride))
        self_idx = b.select(in_bounds, tid, b.const_i32(0))
        left_idx = b.select(do_add, b.sub(tid, c_stride), b.const_i32(0))
        self_vec = b.smem_load_vN(lds_buf, self_idx, dtype=I32, n=1)
        left_vec = b.smem_load_vN(lds_buf, left_idx, dtype=I32, n=1)
        self_val = b.vec_extract(self_vec, 0)
        left_val = b.vec_extract(left_vec, 0)
        new_val = b.add(self_val, left_val)
        b.sync()
        with b.scf_if(do_add):
            b.smem_store_vN(lds_buf, [tid], new_val, 1)
        b.sync()
        stride *= 2

    # Convert inclusive -> exclusive via a one-position right-shift:
    # ``lane k`` writes the value previously at ``lane k - 1`` (or 0
    # for ``lane 0``). A two-phase shift (read all, sync, write all)
    # avoids a same-cycle read/write race on the same LDS slot.
    in_range_left = b.land(in_bounds, b.cmp_gt(tid, b.const_i32(0)))
    # Phase 1: every in-bounds lane materialises its target value in
    # a register. Lanes that aren't in-bounds compute a dummy 0 so the
    # SSA values dominate the phase-2 write; the predicate around the
    # actual store keeps OOB lanes from clobbering anything.
    left_idx = b.select(in_range_left, b.sub(tid, b.const_i32(1)), b.const_i32(0))
    left_vec = b.smem_load_vN(lds_buf, left_idx, dtype=I32, n=1)
    left_val = b.vec_extract(left_vec, 0)
    shifted = b.select(in_range_left, left_val, b.const_i32(0))
    b.sync()
    # Phase 2: predicated write-back. The tid==0 slot gets 0; all
    # other in-bounds slots get the value previously held one position
    # to the left -- which is the canonical exclusive scan result.
    with b.scf_if(in_bounds):
        b.smem_store_vN(lds_buf, [tid], shifted, 1)
    b.sync()


def block_two_level_scan_i32(
    b: IRBuilder,
    lds_buf: Value,
    *,
    tid: Value,
    block_size: int,
    length: int,
    wave_size: int = 64,
) -> None:
    """Two-level exclusive prefix-sum: per-wave Kogge-Stone + cross-wave merge.

    For ``length > 64`` (the moe_sorting target with ``E in {64, 128,
    256}``), the canonical :func:`block_exclusive_scan_i32` issues
    ``log2(length)`` LDS round-trips, each guarded by an
    ``s_barrier``. P34 promotes AITER's ``moe_sorting_opus.h::
    wave_cumsum`` shape: each lane in a wave does a 6-stage
    Kogge-Stone scan via wave-internal shuffles (no LDS, no
    barrier), the per-wave totals land in a ``num_warps`` LDS
    slot table, ``wave 0`` scans the totals (≤4 stages), and the
    cross-wave prefix is added back to each wave's local prefix.

    Halves LDS round-trip count from 7 to 1 for ``length=128`` and
    saves the matching number of barriers — biggest absolute win in
    the persistent / fused MP variant of moe_sorting (P63) where the
    scan is hot.

    The implementation falls back to the existing single-level
    Hillis-Steele scan when ``length <= wave_size`` (it's already
    optimal in that regime) or when the kernel ``block_size`` cannot
    be cleanly divided into ``wave_size`` waves.
    """
    if length <= 0:
        raise ValueError(f"length must be > 0 (got {length})")
    # Single-wave fall-back: the existing Hillis-Steele scan is already
    # at the lower bound for ``length <= wave_size``.
    if length <= wave_size or block_size % wave_size != 0:
        block_exclusive_scan_i32(
            b, lds_buf, tid=tid, block_size=block_size, length=length
        )
        return
    if length > block_size:
        raise ValueError(
            f"block_two_level_scan_i32: length {length} > block_size "
            f"{block_size}; multi-pass scans not implemented yet"
        )

    # Stage 1: per-wave Kogge-Stone. Each wave handles ``wave_size``
    # consecutive slots; lane ``l`` in wave ``w`` reads ``lds_buf[w *
    # wave_size + l]`` from the input and produces an in-wave inclusive
    # scan via :meth:`IRBuilder.warp_shuffle_xor` shuffles. The
    # inclusive form is converted to exclusive by a one-lane right
    # shift via ``warp_shuffle_xor`` with mask 1 + a lane-0 zero
    # substitute.
    num_waves = length // wave_size
    if length % wave_size != 0:
        # Tail: residual `length % wave_size` lanes in the last wave
        # — we keep the canonical scan for the residual to avoid a
        # second tier of arithmetic. Falls back to the single-level
        # scan when length is not a clean multiple of wave_size.
        block_exclusive_scan_i32(
            b, lds_buf, tid=tid, block_size=block_size, length=length
        )
        return

    c_wave = b.const_i32(wave_size)
    lane = b.mod(tid, c_wave)
    warp = b.div(tid, c_wave)
    in_bounds = b.cmp_lt(tid, b.const_i32(length))

    # Load my slot's input value.
    self_idx = b.select(in_bounds, tid, b.const_i32(0))
    self_v = b.vec_extract(b.smem_load_vN(lds_buf, self_idx, dtype=I32, n=1), 0)
    b.sync()

    # Per-wave Kogge-Stone inclusive scan via warp_shuffle_xor.
    # AMDGPU's ds_swizzle handles the within-wave shuffle in 1 LDS op
    # (vs ds_bpermute's address-arithmetic + LDS round-trip).
    cur = self_v
    stages = wave_size.bit_length() - 1
    for k in range(stages):
        mask = 1 << k
        remote = b.warp_shuffle_xor(cur, mask)
        # Lanes whose XOR neighbour is "earlier" (i.e. lane ^ mask
        # < lane) add the remote partial; the others see no change.
        is_higher = b.cmp_ge(lane, b.const_i32(mask))
        cur = b.select(is_higher, b.add(cur, remote), cur)

    inclusive = cur

    # Per-wave totals: lane (wave_size - 1) holds the wave-local sum.
    # Stash to LDS so wave 0 can scan them.
    is_wave_tail = b.cmp_eq(lane, b.const_i32(wave_size - 1))
    with b.scf_if(b.land(is_wave_tail, in_bounds)):
        b.smem_store_vN(lds_buf, [warp], inclusive, 1)
    b.sync()

    # Cross-wave scan: wave 0 reads num_waves entries, scans them
    # serially (small N: typically 2 / 4 / 8 — fewer cycles than another
    # round of warp shuffles), and writes back per-wave prefixes.
    is_wave0 = b.cmp_eq(warp, b.const_i32(0))
    is_xfer_lane = b.land(is_wave0, b.cmp_lt(lane, b.const_i32(num_waves)))
    with b.scf_if(is_xfer_lane):
        # Read every wave's total into the lane-0 wave.
        my_total = b.vec_extract(b.smem_load_vN(lds_buf, lane, dtype=I32, n=1), 0)
        # Synthesise the cross-wave exclusive prefix by reading lower
        # waves' totals and summing. ``num_waves`` is small so this
        # unrolled loop costs at most 7 adds.
        prefix = b.const_i32(0)
        for w_lower in range(num_waves):
            other_total = b.vec_extract(
                b.smem_load_vN(lds_buf, b.const_i32(w_lower), dtype=I32, n=1),
                0,
            )
            include_lower = b.cmp_lt(b.const_i32(w_lower), lane)
            prefix = b.select(include_lower, b.add(prefix, other_total), prefix)
        del my_total  # informational; the in-wave scan recomputes it
        b.smem_store_vN(lds_buf, [lane], prefix, 1)
    b.sync()

    # Add the cross-wave prefix back into each lane's inclusive scan,
    # then shift right by 1 to make the scan exclusive.
    cross_prefix = b.vec_extract(b.smem_load_vN(lds_buf, warp, dtype=I32, n=1), 0)
    full_inclusive = b.add(inclusive, cross_prefix)
    b.sync()

    # Inclusive -> exclusive: lane writes its left neighbour's
    # inclusive value (or 0 for lane 0 of wave 0).
    in_range_left = b.land(in_bounds, b.cmp_gt(tid, b.const_i32(0)))
    # Stash inclusive scan back into LDS so the right-shift can read it.
    with b.scf_if(in_bounds):
        b.smem_store_vN(lds_buf, [tid], full_inclusive, 1)
    b.sync()
    left_idx = b.select(in_range_left, b.sub(tid, b.const_i32(1)), b.const_i32(0))
    left_vec = b.smem_load_vN(lds_buf, left_idx, dtype=I32, n=1)
    shifted = b.select(in_range_left, b.vec_extract(left_vec, 0), b.const_i32(0))
    b.sync()
    with b.scf_if(in_bounds):
        b.smem_store_vN(lds_buf, [tid], shifted, 1)
    b.sync()

File: dnn-providers/hip-kernel-provider/rocke/platform/Cpp/helpers/scan.cpp
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT
/*
 * helper_rocke.helpers.scan.c -- C99 port of selected symbols from
 * rocke/helpers/scan.py.
 *
 * Ported symbols: lds_zero_i32, block_exclusive_scan_i32.
 *
 * Each helper reproduces its Python counterpart's rocke_b_* builder-call sequence
 * byte-faithfully (same ops, same order, same operands). The host-side control
 * structure (the chunk loop, the Hillis-Steele while-loop, the scf_if scoping)
 * is reproduced exactly so the emitted op stream is identical to the Python.
 *
 * scf_if scoping: Python's ``with b.scf_if(cond):`` maps to
 *   rocke_if_t iff = rocke_b_scf_if(b, cond);
 *   rocke_b_region_enter(b, iff.then_region);
 *   ... body ...
 *   rocke_b_region_leave(b);
 *
 * Lifetime: every node is arena-owned (rocke_ir_builder_t.arena). Nothing is freed
 * individually; the arena bulk-frees the whole graph.
 */

#include <stdarg.h>
#include <stdint.h>
#include <stdio.h>

#include "rocke/error.hpp"
#include "rocke/helper_rocke.helpers.scan.h"
#include "rocke/ir.h"

/* ----------------------------------------------------------------- helpers */

/* Raise the failure as a ckc::Error (mirroring the Python `raise`); the public
 * entry boundary catches it and records status + message on the builder, so the
 * C ABI is unchanged. [[noreturn]] keeps the existing `rocke_scan_set_err(...);
 * return;` call sites valid -- the trailing return is simply never reached. */
[[noreturn]] static void
    rocke_scan_set_err(rocke_ir_builder_t* b, rocke_status_t st, const char* fmt, ...)
{
    (void)b;
    char msg[ROCKE_ERR_MSG_CAP];
    va_list ap;
    va_start(ap, fmt);
    (void)vsnprintf(msg, sizeof(msg), fmt, ap);
    va_end(ap);
    msg[sizeof(msg) - 1] = '\0';
    ckc::raise_status(st, msg);
}

/* ------------------------------------------------------------- lds_zero_i32 */

void rocke_lds_zero_i32(
    rocke_ir_builder_t* b, rocke_value_t* lds_buf, rocke_value_t* tid, int block_size, int length)
{
    int chunks;
    int c;
    rocke_value_t* c_block;
    rocke_value_t* c_length;
    rocke_value_t* c_zero;

    if(b == NULL)
    {
        return;
    }
    if(length <= 0)
    {
        rocke_scan_set_err(b, ROCKE_ERR_VALUE, "length must be > 0 (got %d)", length);
        return;
    }

    chunks = (length + block_size - 1) / block_size;
    c_block = rocke_b_const_i32(b, block_size);
    c_length = rocke_b_const_i32(b, length);
    c_zero = rocke_b_const_i32(b, 0);

    for(c = 0; c < chunks; ++c)
    {
        /* local = b.add(tid, b.mul(b.const_i32(c), c_block)) */
        rocke_value_t* local
            = rocke_b_add(b, tid, rocke_b_mul(b, rocke_b_const_i32(b, c), c_block));
        /* in_bounds = b.cmp_lt(local, c_length) */
        rocke_value_t* in_bounds = rocke_b_cmp_lt(b, local, c_length);
        /* with b.scf_if(in_bounds): b.smem_store_vN(lds_buf, [local], c_zero, 1) */
        rocke_if_t iff = rocke_b_scf_if(b, in_bounds);
        rocke_b_region_enter(b, iff.then_region);
        {
            rocke_value_t* idx[1];
            idx[0] = local;
            rocke_b_smem_store_vN(b, lds_buf, idx, 1, c_zero, 1);
        }
        rocke_b_region_leave(b);
    }
    rocke_b_sync(b);
}

/* ------------------------------------------------- block_exclusive_scan_i32 */

void rocke_block_exclusive_scan_i32(
    rocke_ir_builder_t* b, rocke_value_t* lds_buf, rocke_value_t* tid, int block_size, int length)
{
    const rocke_type_t* I32;
    rocke_value_t* c_length;
    rocke_value_t* in_bounds;
    int stride;
    rocke_value_t* in_range_left;
    rocke_value_t* left_idx;
    rocke_value_t* left_vec;
    rocke_value_t* left_val;
    rocke_value_t* shifted;
    rocke_if_t iff_final;

    if(b == NULL)
    {
        return;
    }
    if(length <= 0)
    {
        rocke_scan_set_err(b, ROCKE_ERR_VALUE, "length must be > 0 (got %d)", length);
        return;
    }
    if(length > block_size)
    {
        rocke_scan_set_err(b,
                           ROCKE_ERR_VALUE,
                           "length %d > block_size %d; multi-pass scans not implemented yet",
                           length,
                           block_size);
        return;
    }

    I32 = rocke_i32();

    /* c_length = b.const_i32(length); in_bounds = b.cmp_lt(tid, c_length) */
    c_length = rocke_b_const_i32(b, length);
    in_bounds = rocke_b_cmp_lt(b, tid, c_length);

    /* Inclusive Hillis-Steele scan. */
    stride = 1;
    while(stride < length)
    {
        /* c_stride = b.const_i32(stride) */
        rocke_value_t* c_stride = rocke_b_const_i32(b, stride);
        /* do_add = b.land(in_bounds, b.cmp_ge(tid, c_stride)) */
        rocke_value_t* do_add = rocke_b_land(b, in_bounds, rocke_b_cmp_ge(b, tid, c_stride));
        /* self_idx = b.select(in_bounds, tid, b.const_i32(0)) */
        rocke_value_t* self_idx = rocke_b_select(b, in_bounds, tid, rocke_b_const_i32(b, 0));
        /* left_idx = b.select(do_add, b.sub(tid, c_stride), b.const_i32(0)) */
        rocke_value_t* l_idx
            = rocke_b_select(b, do_add, rocke_b_sub(b, tid, c_stride), rocke_b_const_i32(b, 0));
        /* self_vec = b.smem_load_vN(lds_buf, self_idx, dtype=I32, n=1) */
        rocke_value_t* self_vec;
        rocke_value_t* l_vec;
        rocke_value_t* self_val;
        rocke_value_t* l_val;
        rocke_value_t* new_val;
        rocke_if_t iff;
        rocke_value_t* sidx[1];
        rocke_value_t* lidx[1];
        rocke_value_t* widx[1];

        sidx[0] = self_idx;
        self_vec = rocke_b_smem_load_vN(b, lds_buf, sidx, 1, I32, 1);
        /* left_vec = b.smem_load_vN(lds_buf, left_idx, dtype=I32, n=1) */
        lidx[0] = l_idx;
        l_vec = rocke_b_smem_load_vN(b, lds_buf, lidx, 1, I32, 1);
        /* self_val = b.vec_extract(self_vec, 0); left_val = b.vec_extract(left_vec, 0) */
        self_val = rocke_b_vec_extract(b, self_vec, 0);
        l_val = rocke_b_vec_extract(b, l_vec, 0);
        /* new_val = b.add(self_val, left_val) */
        new_val = rocke_b_add(b, self_val, l_val);
        /* b.sync() */
        rocke_b_sync(b);
        /* with b.scf_if(do_add): b.smem_store_vN(lds_buf, [tid], new_val, 1) */
        iff = rocke_b_scf_if(b, do_add);
        rocke_b_region_enter(b, iff.then_region);
        {
            widx[0] = tid;
            rocke_b_smem_store_vN(b, lds_buf, widx, 1, new_val, 1);
        }
        rocke_b_region_leave(b);
        /* b.sync() */
        rocke_b_sync(b);

        stride *= 2;
    }

    /* Convert inclusive -> exclusive via a one-position right-shift. */
    /* in_range_left = b.land(in_bounds, b.cmp_gt(tid, b.const_i32(0))) */
    in_range_left = rocke_b_land(b, in_bounds, rocke_b_cmp_gt(b, tid, rocke_b_const_i32(b, 0)));
    /* left_idx = b.select(in_range_left, b.sub(tid, b.const_i32(1)), b.const_i32(0)) */
    left_idx = rocke_b_select(
        b, in_range_left, rocke_b_sub(b, tid, rocke_b_const_i32(b, 1)), rocke_b_const_i32(b, 0));
    /* left_vec = b.smem_load_vN(lds_buf, left_idx, dtype=I32, n=1) */
    {
        rocke_value_t* lidx2[1];
        lidx2[0] = left_idx;
        left_vec = rocke_b_smem_load_vN(b, lds_buf, lidx2, 1, I32, 1);
    }
    /* left_val = b.vec_extract(left_vec, 0) */
    left_val = rocke_b_vec_extract(b, left_vec, 0);
    /* shifted = b.select(in_range_left, left_val, b.const_i32(0)) */
    shifted = rocke_b_select(b, in_range_left, left_val, rocke_b_const_i32(b, 0));
    /* b.sync() */
    rocke_b_sync(b);
    /* with b.scf_if(in_bounds): b.smem_store_vN(lds_buf, [tid], shifted, 1) */
    iff_final = rocke_b_scf_if(b, in_bounds);
    rocke_b_region_enter(b, iff_final.then_region);
    {
        rocke_value_t* widx2[1];
        widx2[0] = tid;
        rocke_b_smem_store_vN(b, lds_buf, widx2, 1, shifted, 1);
    }
    rocke_b_region_leave(b);
    /* b.sync() */
    rocke_b_sync(b);
}
